Extract remove_worker logic from MiqServer

kbrock opened the pull request:
https://github.com/ManageIQ/manageiq/issues/21716

This is focused on the "killing" of our MiqServer#miq_worker and our in memory store of them.

simplify the way that we traverse the various workers that we are going to stop/kill/quiesce.

CHANGE: cut down the scope of this to just focus around the miq_worker

# Changes in the code

We had 2 code patterns for cleaning out the lists:
- `worker.destroy`, `miq_workers.reload`
- `miq_workers.delete(workers)`

If there is a reason to use one of the other, please share

Fryguy commented on app/models/miq_server/worker_management/monitor.rb:
```diff
@@ -52,16 +52,12 @@ def cleanup_failed_workers
+
+    remove_workers(miq_workers.to_a.select(&:is_stopped?)) do |w|
       _log.info("SQL Record for #{w.format_full_log_msg}, Status: [#{w.status}] is being deleted")
-      worker_delete(w.pid)
```
But then it doesn't delete any?

kbrock replied:
`remove_workers` does that for us.
I debated removing this logging but this PR was aiming for a refactor only. (at least so far)

Fryguy commented:
I know it's still WIP, but there is too many disparate things going on in this PR that it's confusing to follow.  The smaller refactors like renaming methods (e.g. renaming `remove_unknown_workers` to `kill_unknown_worker_processes`) should probably move out to separate PRs otherwise they make it really hard to see what's going on.

kbrock commented:
push: cops

kbrock commented:
@Fryguy refactored this WIP PR and cut down the scope a lot.

If you go through the commits, I believe it shows the rename and the refactor. Let me know if you want me to put more commits into there.

I also put the high level goals and the steps into the issue

Fryguy commented:
This refactor is a lot clearer now.  From what I can see, this LGTM.  @jrafanie and/or @agrare Please also review.

agrare commented:
:+1: looks good @kbrock

kbrock commented:
@Fryguy Is this a good size for a PR?
I have at least 3x the amount of these changes but am trying to hold off so it can fit in the mind best as possible

Fryguy commented:
Yes, this is perfect.  It's very clear here with a singular goal of DRYing up this pattern:

```
    workers.each do |w|
      yield(w)
      worker_delete(w.pid)
      w.destroy
    end
```

Fryguy commented on app/models/miq_server/worker_management/monitor/stop.rb:
```diff
@@ -14,12 +14,8 @@ def stop_worker_queue(worker, monitor_reason = nil)
-    w = worker.kind_of?(Integer) ? miq_workers.find_by(:id => worker) : worker
-
-    if w.nil?
-      _log.warn("Cannot find Worker <#{w.inspect}>")
```
Noticed this was removed.  Any particular reason or is it just not useful?

kbrock replied:
This log entry was pulled out into `find_worker`.
I wanted to reuse this when I transitioned 'kill_async` over to be more similar to `stop_worker_queue`.

Maybe the log should not be hidden in a finder method and I should drop this commit

Fryguy commented:
@jrafanie Please also review.

kbrock commented on app/models/miq_server/worker_management/monitor/kill.rb:
```diff
@@ -19,22 +19,16 @@ def remove_unknown_workers
-      killed_workers << w
+  def kill_worker_processes(workers = miq_workers)
+    remove_workers(workers) do |w|
+      w.kill_process if w.current_or_starting?
```
This would be simplified if we only wanted to remove the records that were `current_or_starting?`

instead:
-  kill the `killing_or_starting?`
- remove every miq worker from the miq worker list.

kbrock commented on app/models/miq_server/worker_management/process.rb:
```diff
@@ -38,18 +38,14 @@ def workers_quiesced?
-    miq_workers.each do |w|
-      next unless monitor_reason_not_responding?(w)
-      next unless worker_get_monitor_status(w.pid) == :waiting_for_stop
+    workers_to_check = miq_workers.select do |w|
+      monitor_reason_not_responding?(w) &&
+        worker_get_monitor_status(w.pid) == :waiting_for_stop
+    end
```
in contrast, this is only working on the sub select of records (calls a `next` otherwise). So converting over to a `select(&:)` makes sense - and I find it easier to read and reason.

kbrock commented on app/models/miq_server/worker_management.rb:
```diff
@@ -71,4 +71,17 @@ def worker_add(worker_pid)
+      worker_delete(w.pid)
+      w.destroy
+    end
+    miq_workers.delete(*workers)
```
For the removal of `miq_workers` we had 2 patterns.

They seemed similar enough so I just picked one. If you like the other one, then I can swap to that.

### reload pattern

You can find an example in [process.rb:41](https://github.com/ManageIQ/manageiq/blob/3fb1aa56ccf5b85f4893e5fa6ada0937a27377cf/app/models/miq_server/worker_management/process.rb#L41-L52)

```ruby
    worker_deleted = false
    miq_workers.each do |w|
      # [...]
      w.destroy
      worker_deleted = true
    end
    miq_workers.reload if worker_deleted
```

### delete pattern

This example is in [kill.rb](https://github.com/ManageIQ/manageiq/blob/3fb1aa56ccf5b85f4893e5fa6ada0937a27377cf/app/models/miq_server/worker_management/monitor/kill.rb#L30-L38)

```ruby
    killed_workers = []
    workers.each do |w|
      # [...]
      w.destroy
      killed_workers << w
    end
    miq_workers.delete(*killed_workers) unless killed_workers.empty?
```

chessbyte replied:
I prefer the reload pattern

jrafanie replied:
My bad.  I was confusing `miq_workers.delete` here with the `worker_delete(w.pid)` to update the in memory cache.

I think this is the same logic.

kbrock replied:
I had liked the delete pattern, but after looking at those 2 examples, the reload speaks to me more.
I went ahead and changed that

jrafanie replied:
Thanks, I like it.  It's more explicit and you won't make the mistake I did in transposing the `miq_workers.delete` and the `worker_delete`.

kbrock commented:
update:
- switched from `collection.delete` to `collection.reload`. Since we are deleting most of the records anyway, refresh will bring back very few and may pickup blips a little better.
- rebased
